src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/HttpClientFactoryDisposeTests.cs

```diff
@@ -0,0 +1,148 @@
+                handlers.Add(handler);
+            }
+            
+            // Verify handlers were created
+            Assert.Equal(ClientCount, disposeCounter.Created);
+            
+            // Act - dispose the service provider and handlers directly
+            serviceProvider.Dispose();
+            foreach (var handler in handlers)
+            {
+                handler.Dispose();
+            }
+            
+            // Assert - all handlers should be disposed
+            Assert.Equal(disposeCounter.Created, disposeCounter.Disposed);
+        }
+
+        private class DisposeCounter
+        {
+            private int _created;
+            private int _disposed;
+
+            public int Created => Interlocked.CompareExchange(ref _created, 0, 0);
+            public int Disposed => Interlocked.CompareExchange(ref _disposed, 0, 0);
+
+            public void IncrementCreated()
+            {
+                Interlocked.Increment(ref _created);
+            }
+
+            public void IncrementDisposed()
+            {
+                Interlocked.Increment(ref _disposed);
+            }
+        }
+
+        private class DisposeTrackingHandler : HttpMessageHandler
+        {
+            private readonly DisposeCounter _counter;
+
+            public DisposeTrackingHandler(DisposeCounter counter)
+            {
+                _counter = counter;
+                _counter.IncrementCreated();
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                if (disposing)
+                {
+                    _counter.IncrementDisposed();
+                }
+
+                base.Dispose(disposing);
+            }
+
+            protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                // Just return a simple response for test purposes
+                return Task.FromResult(new HttpResponseMessage(System.Net.HttpStatusCode.OK));
+            }
+        }
+
+        private class TestHttpClientFactory : DefaultHttpClientFactory
+        {
+            private readonly IServiceProvider _services;
+            private readonly IServiceScopeFactory _scopeFactory;
+            private readonly DisposeCounter _counter;
+
+            public TestHttpClientFactory(
+                IServiceProvider services,
+                IServiceScopeFactory scopeFactory,
+                IOptionsMonitor<HttpClientFactoryOptions> optionsMonitor,
+                IEnumerable<IHttpMessageHandlerBuilderFilter> filters)
+                : base(services, scopeFactory, optionsMonitor, filters)
+            {
+                _services = services;
+                _scopeFactory = scopeFactory;
+                _counter = services.GetRequiredService<DisposeCounter>();
+            }
+
+            public void CreateHandlersForTesting(int count)
```

Comment:
I don't think it's a good idea to mangle with the internals of the factory directly.
A better approach is to use something like https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/DefaultHttpClientFactoryTest.cs#L536

then you can:
- create N handlers "normally" by asking for httpclients
- use the TCS to force expiry -> now we have N expired handlers
- create M more handlers "normally". now we also have M active handlers

you can take a look at the tests in that file.